python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,533 @@
+            # 3. module output to moudle input
+            """
+            if self.io_owner == binding.io_owner:
+                raise RuntimeError(f"can not set self as binding.")
+
+            if not self.is_global_interface() and self.io_type == "input":
+                raise RuntimeError(f"Module only can start binding from output!")
+
+            if (
+                not self.is_global_interface()
+                and not binding.is_global_interface()
+                and binding.io_type == "output"
+            ):
+                raise RuntimeError(f"Module output can not binding with module output!")
+
+            if (
+                not self.is_global_interface()
+                and binding.is_global_interface()
+                and binding.io_type == "input"
+            ):
+                raise RuntimeError(f"Module output can not binding with global input!")
+
+            if self.is_global_interface() and self.io_type != "input":
+                raise RuntimeError(f"Global only can start binding from input!")
+
+            if self.is_global_interface() and binding.io_type != "input":
+                raise RuntimeError(f"Global input only can set binding with module input.")
+
+            self.bindings.append(binding)
+            if not self.is_global_interface():
+                # check if the source and target data_type same
+                if (
+                    isinstance(binding.io_owner, PipelineConfig.ModuleWrapper)
+                    and self.data_type != binding.data_type
+                ):
+                    raise RuntimeError(
+                        f"Illegal type (%s vs. %s): binding type is not same!"
+                        % (self.data_type, binding.data_type)
+                    )
+
+                binding.parents.append(self)
+                # Do acyclic check after increase the in-degree.
+                if not self.check_dag_acyclic(
+                    binding.io_owner, self.io_owner.input_bindings.bindings
+                ):
+                    raise RuntimeError(f"Illegal connection: cause a circle!")
+
+    class BindingList:
+        """Container for bindings(input or output interface).
+
+        Parameters
+        ----------
+        owner : ModuleWrapper/PipelineConfig
+            The owner of this list, it can be ModuleWrapper or PipelineConfig
+
+        type_name : str
+            The type of this binding list. It can be either "input" or "output".
+        """
+
+        def __init__(self, owner, type_name):
+            self.bindings = {}
+            self.io_owner = owner
+            self.binding_type = type_name
+
+        def get_binding_data_type(self, key):
+            """return binding data type"""
+            if isinstance(self.io_owner, PipelineConfig.ModuleWrapper):
+                return self.io_owner.get_data_type(key, self.binding_type)
+            return None
+
+        def __getitem__(self, key):
+            """return item by key"""
```

Comment:
Remove this comment as it is not helpful.

Reply:
removed.